gcp_deepvariant_runner.py

```diff
@@ -321,6 +395,83 @@ def _get_gcs_relative_path(gcs_path):
+    logging.error('Size of input bam file is 0.')
+
+  is_wes = pipeline_args.model.find(_WES_MODEL_IDENTIFIER) != -1
+  is_wgs = pipeline_args.model.find(_WGS_MODEL_IDENTIFIER) != -1
+
+  if is_wes:
+    if bam_size < _WES_LARGE_THR:
+      return BamCategories.WES_SMALL
+    else:
+      return BamCategories.WES_LARGE
+
+  if is_wgs:
+    if bam_size < _WGS_SMALL_THR:
+      return BamCategories.WGS_SMALL
+    elif bam_size > _WGS_LARGE_THR:
+      return BamCategories.WGS_LARGE
+    else:
+      return BamCategories.WGS_MEDIUM
+
+
+def _set_args_using_optimized_default_flags(default_flags, pipeline_args):
+  """Sets pipeline args using the set of given default flag values."""
+  pipeline_args.shards = (
+      default_flags['make_examples_workers'] *
+      default_flags['make_examples_cores_per_worker'])
+  pipeline_args.make_examples_workers = (default_flags['make_examples_workers'])
+  pipeline_args.make_examples_cores_per_worker = (
+      default_flags['make_examples_cores_per_worker'])
+  pipeline_args.make_examples_ram_per_worker_gb = (
+      default_flags['make_examples_cores_per_worker'] * _RAM_PER_CORE)
+  pipeline_args.make_examples_disk_per_worker_gb = (
+      _MAKE_EXAMPLES_DISK_PER_WORKER)
+  if 'gpu' in default_flags:
+    pipeline_args.gpu = default_flags['gpu']
+    pipeline_args.call_variants_workers = (
+        default_flags['call_variants_workers'])
+    pipeline_args.call_variants_cores_per_worker = (
+        default_flags['call_variants_cores_per_worker'])
+    pipeline_args.call_variants_ram_per_worker_gb = (
+        default_flags['call_variants_cores_per_worker'] * _RAM_PER_CORE)
+    pipeline_args.call_variants_disk_per_worker_gb = (
+        _CALL_VARIANTS_DISK_PER_WORKER)
+  elif 'tpu' in default_flags:
+    pipeline_args.tpu = default_flags['tpu']
+    pipeline_args.gke_cluster_zone = pipeline_args.zones[0]
+  else:
+    raise ValueError('Either gpu or tpu is needed for default flag settings.')
+  # Following flags are independent of BAM file category.
+  pipeline_args.preemptible = True
+  if pipeline_args.gvcf_outfile:
+    pipeline_args.postprocess_variants_disk_gb = _POSTPROCESS_VARIANTS_DISK_GVCF
+
+
+def _set_computational_flags_based_on_bam_size(pipeline_args):
+  """Automatically sets computational flags based on size of input BAM file."""
+  # First validating all necessary flags are present.
+  if not (pipeline_args.docker_image and pipeline_args.docker_image_gpu):
+    raise ValueError('both --docker_image and --docker_image_gpu must be '
+                     'provided with --set_optimized_flags_based_on_bam_size')
+  is_wes = pipeline_args.model.find(_WES_MODEL_IDENTIFIER) != -1
+  is_wgs = pipeline_args.model.find(_WGS_MODEL_IDENTIFIER) != -1
+  if is_wes == is_wgs:
+    raise ValueError('Unable to automatically set computational flags. Given '
+                     'model is neither WGS nor WES: %s' % pipeline_args.model)
+  if not pipeline_args.bam.endswith(_BAM_FILE_SUFFIX):
```

Comment:
Why cannot have it for CRAM? Is it not yet supported or it's not possible?

Anyway, this check doesn't belong to here best. Please move it to where flags are validated.

Reply:
Since CRAM files might have different levels of compression, we cannot simply rely on their size. So for now I decided to limit this to BAM files.

And for the placement, as I mentioned in the previous comment, here we check everything related to automatic flag values prior to setting them. Please let me know if this makes sense.